internal/report/report.go

```diff
@@ -0,0 +1,271 @@
+	// This will open every file normally expect for files that have been modified by Orchestrion, in which case it will
+	// open the modified file instead of the original file.
+	Report struct {
+		files []ModifiedFile
+		fs    fs.FS
+		root  string
+	}
+)
+
+// NewModifiedFile creates a new ModifiedFile from the given file system and modified path.
+// If the modified file contains a line directive, it will parse it to find the original path.
+func NewModifiedFile(fsys fs.FS, modifiedPath string) (ModifiedFile, error) {
+	modifiedFile, err := fsys.Open(modifiedPath)
+	if err != nil {
+		return ModifiedFile{}, fmt.Errorf("open %s: %w", modifiedPath, err)
+	}
+
+	defer modifiedFile.Close()
+
+	originalPath, err := parse.ConsumeLineDirective(modifiedFile)
+	if err != nil {
+		return ModifiedFile{}, fmt.Errorf("consume line directive: %w", err)
+	}
+
+	if originalPath != "" {
+		originalPath = filepath.Clean(originalPath)
+	}
+
+	return ModifiedFile{
+		original: originalPath,
+		modified: filepath.Clean(modifiedPath),
+	}, nil
+}
+
+func (m ModifiedFile) String() string {
+	original := m.original
+	if m.original == "" {
+		original = "(unknown original path)"
+	}
+	return fmt.Sprintf("%s -> %s", original, m.modified)
+}
+
+// ImportPath converts the modified file path to an import path.
+func (m ModifiedFile) ImportPath() string {
+	dir := filepath.Dir(m.modified)
+	_, pkg, found := strings.Cut(dir, aspect.OrchestrionDirPathElement)
+	if !found {
+		return ""
+	}
+
+	return strings.Trim(pkg, "/")
+}
+
+var _ fs.FS = (*Report)(nil)
+
+func (r Report) Open(name string) (fs.File, error) {
+	for _, file := range r.files {
+		if file.original == name {
+			name = file.modified
+			break
+		}
+	}
+
+	return r.fs.Open(name)
+}
+
+var _ fs.FS = (*Report)(nil)
+
+// WithRegexFilter filters the files in the report based on a regex pattern.
+func (r Report) WithRegexFilter(regex string) (Report, error) {
+	cmpRegex, err := regexp.Compile(regex)
+	if err != nil {
+		return Report{}, fmt.Errorf("invalid regex %q: %w", regex, err)
+	}
+
+	return Report{
+		files: slices.DeleteFunc(r.files, func(file ModifiedFile) bool {
+			return !cmpRegex.MatchString(file.modified)
+		}),
+	}, nil
+}
+
+// WithSpecialCasesFilter filters the files in the report to include only those that are not weaver special cases
+func (r Report) WithSpecialCasesFilter() Report {
+	return Report{
+		files: slices.DeleteFunc(r.files, func(file ModifiedFile) bool {
+			pkgPath := file.ImportPath()
+			if pkgPath == "synthetic" {
+				return true
+			}
+			behaviour, isSpecial := aspect.FindBehaviorOverride(pkgPath)
+			return isSpecial && behaviour != aspect.NoOverride
+		}),
+	}
+}
+
+// Diff generates a diff between the original and modified files and writes it to the writer.
+func (r Report) Diff(writer io.Writer) error {
+	var errs []error
+
+	// Check if the diff command is available, exit early if not
+	if _, err := exec.LookPath("diff"); err != nil {
```

Comment:
Maybe in the subsequent PRs, we can consider `git diff` facilities. e.g. I use fancy git diff setup maybe it is possible to use whatever is set for git.